### PR TITLE
Add random_state to seed for onedal preview RF/ET

### PR DESCRIPTION
**Pull request by icfaust**
# Enable setting random_state for RF and ET algos in preview namespace
Random Forest and Extra Trees algos in onedal require random_state to be enabled for deterministic behavior. This is important for performance evaluation and for proper conformance to testing.

Changes proposed in this pull request:
- add random_state -> seed in onedal/ensemble/forest.py
- remove usage of check_random_state to a value in any fit function in sklearnex/preview/ensemble/forest.py
- remove usage of check_random_state to a value in any fit function in sklearnex/preview/ensemble/extra_trees.py
- disable associated warnings in sklearnex/preview/ensemble/

### CODE REVIEW

File: onedal/ensemble/forest.py [hunk 1]
```diff
@@ -206,6 +206,9 @@ def _get_onedal_params(self, data):
                 ceil(
                     self.min_samples_split * n_samples)))
 
+        rs_ = check_random_state(self.random_state)
+        seed_ = rs_.randint(0, np.iinfo("i").max)
```
Comment (KulikovNikita) [inline onedal/ensemble/forest.py:210]: What is the point of using underscored variable names in the internal functionality?

Comment (icfaust) [reply onedal/ensemble/forest.py:210]: Good eye, I was following similar conventions that was in daal4py, but its not even a good python convention, the underscore should be before the variable to indicate private.

File: onedal/ensemble/forest.py [hunk 2]
```diff
@@ -206,6 +206,9 @@ def _get_onedal_params(self, data):
                 ceil(
                     self.min_samples_split * n_samples)))
 
+        rs = check_random_state(self.random_state)
+        seed = rs.randint(0, np.iinfo("i").max)
```
Comment (ahuber21) [inline onedal/ensemble/forest.py:210]: Isn't this redundant? What happens if we don't provide a seed? Naively I would expect that it picks a random one as default. What is the behavior of `check_random_state()` if something is wrong?
If it throws, fine, but if not, you may have to add some error handling here.

File: sklearnex/preview/ensemble/extra_trees.py [hunk 3]
```diff
@@ -569,7 +569,8 @@ def _onedal_cpu_supported(self, method_name, *data):
 
             dal_ready = dal_ready and not hasattr(self, 'estimators_')
 
-            if dal_ready and (self.random_state is not None):
+            if dal_ready and (self.random_state is not None) and \
```
Comment (ahuber21) [inline sklearnex/preview/ensemble/extra_trees.py:572]: Why no `check_random_state()` here?

Comment (ahuber21) [reply sklearnex/preview/ensemble/extra_trees.py:572]: Also: I just don't like backslashes https://peps.python.org/pep-0008/#multiline-if-statements

File: sklearnex/preview/ensemble/extra_trees.py [hunk 4]
```diff
@@ -1148,7 +1152,6 @@ def _onedal_fit(self, X, y, sample_weight=None, queue=None):
         self.n_features_in_ = X.shape[1]
         if not sklearn_check_version('1.0'):
             self.n_features_ = self.n_features_in_
-        rs_ = check_random_state(self.random_state)
```
Comment (ahuber21) [inline sklearnex/preview/ensemble/extra_trees.py:1151]: Seeing this here I would expect you want to add a `check_random_state(self.random_state)` above all the `if` statements.

File: sklearnex/preview/ensemble/forest.py [hunk 5]
```diff
@@ -1056,7 +1055,7 @@ def _onedal_fit(self, X, y, sample_weight=None, queue=None):
             'bootstrap': self.bootstrap,
             'oob_score': self.oob_score,
             'n_jobs': self.n_jobs,
-            'random_state': rs_,
+            'random_state': self.random_state,
```
Comment (ahuber21) [inline sklearnex/preview/ensemble/forest.py:1058]: Same question about `check_random_state()` as in the other files.

**Review by ahuber21 — CHANGES_REQUESTED** [review]

**Comment by Alexsandruss** [comment]
@Mergifyio rebase

File: onedal/ensemble/forest.py (hunk 2 above)
Comment (icfaust) [reply onedal/ensemble/forest.py:210]: This matches the daal4py code, check_random_state comes from sklearn internals which follow similar behavior. (https://github.com/intel/scikit-learn-intelex/blob/master/daal4py/sklearn/ensemble/_forest.py#L689)

File: sklearnex/preview/ensemble/extra_trees.py (hunk 4 above)
Comment (icfaust) [reply sklearnex/preview/ensemble/extra_trees.py:1151]: Removing this here was necessary as it is now handled in onedal/ensemble/forest.py. Otherwise we do it twice.

File: sklearnex/preview/ensemble/forest.py (hunk 5 above)
Comment (icfaust) [reply sklearnex/preview/ensemble/forest.py:1058]: Same as above, now centralized in onedal/ensemble/forest.py

File: onedal/ensemble/forest.cpp [hunk 6]
```diff
@@ -184,11 +184,14 @@ struct params2desc {
                         .set_max_leaf_nodes(params["max_leaf_nodes"].cast<std::int64_t>())
                         .set_max_bins(params["max_bins"].cast<std::int64_t>())
                         .set_min_bin_size(params["min_bin_size"].cast<std::int64_t>())
+#if defined(ONEDAL_VERSION) && ONEDAL_VERSION >= 20230300
```
[the remark this answers is not in the document]
Comment (icfaust) [reply onedal/ensemble/forest.cpp:187]: I will do this an push for re-review. This comment smells like a design rule that we should follow @ahuber21 @Alexsandruss @samir-nasibli . Something that should be discussed and/or written down.

**Comment by icfaust** [comment]
Convention for ordering of descriptor follows the ordering from here: https://github.com/oneapi-src/oneDAL/blob/master/cpp/oneapi/dal/algo/decision_forest/common.hpp#L254

**Review by inteldimitrius — APPROVED** [review]
Changes look good to me

**Review by samir-nasibli — APPROVED** [review]
lgtm. Let's wait all CI check.

**Review by Alexsandruss — CHANGES_REQUESTED** [review]

**Comment by Alexsandruss** [comment]
There are failures of trees:
```
FAILED preview/ensemble/tests/test_preview_ensemble.py::test_sklearnex_import_rf_regression
FAILED preview/ensemble/tests/test_preview_ensemble.py::test_sklearnex_import_et_regression
```

File: onedal/ensemble/tests/test_random_forest.py [hunk 7]
```diff
@@ -51,7 +51,7 @@ def test_rf_regression(queue):
 
 
 @pytest.mark.parametrize("queue", get_queues("gpu"))
-def test_rf_regression(queue):
+def test_rf_regression_gpu(queue):
```
Comment (samir-nasibli) [inline onedal/ensemble/tests/test_random_forest.py:54]: Bad design.
If gpu device is not available, you will get empty list for queue. In this case your getting undef behavior at least.
Even passing this test case when gpu device is not available is meaningful. 

Please do not separate device checks between test suits

**Review by samir-nasibli — CHANGES_REQUESTED** [review]

File: onedal/ensemble/tests/test_random_forest.py [hunk 8]
```diff
@@ -38,13 +38,28 @@ def test_rf_classifier(queue):
...
+@pytest.mark.parametrize("queue", get_queues("cpu"))
+def test_rf_regression_cpu(queue):
+    X, y = make_regression(
+        n_samples=100, n_features=4, n_informative=2, random_state=0, shuffle=False
+    )
+    rf = RandomForestRegressor(max_depth=2, random_state=0).fit(X, y, queue=queue)
+    if daal_check_version((2023, "P", 300)):
+        assert_allclose([-6.97], rf.predict([[0, 0, 0, 0]], queue=queue), atol=1e-2)
+    else:
+        assert_allclose([-6.83], rf.predict([[0, 0, 0, 0]], queue=queue), atol=1e-2)
```
Comment (samir-nasibli) [inline onedal/ensemble/tests/test_random_forest.py:50]: you can get device info from the sycl_queue: 
```python
# Insert here your comment why we have different
# check values for different devices and daal versions.
if queue.sycl_device.is_gpu:
    # do your check
else:
    if daal_check_version((2023, "P", 300)):
        # do your check
    else:
        # do your check
```

Please revert your changes with separating and follow branching above.

File: sklearnex/preview/ensemble/tests/test_preview_ensemble.py [hunk 9]
```diff
@@ -44,7 +44,10 @@ def test_sklearnex_import_rf_regression():
     rf = RandomForestRegressor(max_depth=2, random_state=0).fit(X, y)
     assert "sklearnex.preview" in rf.__module__
     pred = rf.predict([[0, 0, 0, 0]])
-    assert_allclose([-6.839], pred, atol=1e-2)
+    if daal_check_version((2023, "P", 300)):
```
Comment (Alexsandruss) [inline sklearnex/preview/ensemble/tests/test_preview_ensemble.py:47]: Formally, it's incorrect: next version is 2024.0

File: onedal/ensemble/tests/test_random_forest.py [hunk 10]
```diff
@@ -74,4 +86,7 @@ def test_rf_regression_random_splitter(queue):
     rf = RandomForestRegressor(max_depth=2, random_state=0, splitter_mode="random").fit(
         X, y, queue=queue
     )
-    assert_allclose([-6.83], rf.predict([[0, 0, 0, 0]], queue=queue), atol=1e-2)
+    if daal_check_version((2023, "P", 300)):
```
Comment (Alexsandruss) [inline onedal/ensemble/tests/test_random_forest.py:89]:
```suggestion
    if daal_check_version((2024, "P", 0)):
```

File: sklearnex/preview/ensemble/tests/test_preview_ensemble.py [hunk 11]
```diff
@@ -44,7 +44,10 @@ def test_sklearnex_import_rf_regression():
     rf = RandomForestRegressor(max_depth=2, random_state=0).fit(X, y)
     assert "sklearnex.preview" in rf.__module__
     pred = rf.predict([[0, 0, 0, 0]])
-    assert_allclose([-6.839], pred, atol=1e-2)
+    if daal_check_version((2024, "P", 000)):
```
Comment (Alexsandruss) [inline sklearnex/preview/ensemble/tests/test_preview_ensemble.py:47]:
```suggestion
    if daal_check_version((2024, "P", 0)):

```

**Comment by samir-nasibli** [comment]
@icfaust please rebase and enable this tests #1393 , if it's required.
